File: avail-js/src/sdk/block.ts
import { ApiPromise } from "@polkadot/api"
import { GenericExtrinsic } from "@polkadot/types"
import { H256, SignedBlock } from "@polkadot/types/interfaces/runtime"
import { fromHexToAscii } from "./utils"
import { EventRecord, Events } from "."

export class Block {
  signedBlock: SignedBlock

  static async New(api: ApiPromise, blockHash: H256): Promise<Block> {
    const block = await api.rpc.chain.getBlock(blockHash)
    return new Block(block)
  }

  static async NewBestBlock(api: ApiPromise): Promise<Block> {
    const blockHash = await api.rpc.chain.getBlockHash()
    return Block.New(api, blockHash)
  }

  static async NewFinalizedBlock(api: ApiPromise): Promise<Block> {
    const blockHash = await api.rpc.chain.getFinalizedHead()
    return Block.New(api, blockHash)
  }

  constructor(block: SignedBlock) {
    this.signedBlock = block
  }

  async blockHash(api: ApiPromise): Promise<H256> {
    return await api.rpc.chain.getBlockHash(this.signedBlock.block.header.number.toNumber())
  }

  async fetchEvents(api: ApiPromise, txIndex?: number): Promise<EventRecord[]> {
    return Events.fetchEvents(api, await this.blockHash(api), txIndex)
  }

  transactionCount(): number {
    return transactionCount(this.signedBlock)
  }

  transactionAll(): GenericExtrinsic[] {
    return transactionAll(this.signedBlock)
  }

  transactionBySigner(signer: string): GenericExtrinsic[] {
    return transactionBySigner(this.signedBlock, signer)
  }

  transactionByIndex(txIndex: number): GenericExtrinsic | null {
    return transactionByIndex(this.signedBlock, txIndex)
  }

  transactionByHash(txHash: H256): GenericExtrinsic[] {
    return transactionByHash(this.signedBlock, txHash)
  }

  transactionByAppId(appId: number): GenericExtrinsic[] {
    return transactionByAppId(this.signedBlock, appId)
  }

  transactionHashToIndex(txHash: H256): number[] {
    return transactionHashToIndex(this.signedBlock, txHash)
  }

  dataSubmissionsCount(): number {
    return dataSubmissionsCount(this.signedBlock)
  }

  dataSubmissionsAll(): DataSubmission[] {
    return dataSubmissionsAll(this.signedBlock)
  }

  dataSubmissionsBySigner(signer: string): DataSubmission[] {
    return dataSubmissionsBySigner(this.signedBlock, signer)
  }

  dataSubmissionsByIndex(txIndex: number): DataSubmission | null {
    return dataSubmissionsByIndex(this.signedBlock, txIndex)
  }

  dataSubmissionsByHash(txHash: H256): DataSubmission | null {
    return dataSubmissionsByHash(this.signedBlock, txHash)
  }

  dataSubmissionsByAppId(appId: number): DataSubmission[] {
    return dataSubmissionsByAppId(this.signedBlock, appId)
  }
}

export function transactionCount(block: SignedBlock): number {
  return block.block.extrinsics.length
}

export function transactionAll(block: SignedBlock): GenericExtrinsic[] {
  return block.block.extrinsics
}

export function transactionBySigner(block: SignedBlock, signer: string): GenericExtrinsic[] {
  return block.block.extrinsics.filter((tx) => {
    return tx.signer.toString() == signer
  })
}

export function transactionByIndex(block: SignedBlock, txIndex: number): GenericExtrinsic | null {
  const transactions = block.block.extrinsics
  if (txIndex >= transactions.length) return null

  return transactions[txIndex]
}

export function transactionByHash(block: SignedBlock, txHash: H256): GenericExtrinsic[] {
  return block.block.extrinsics.filter((tx) => {
    return tx.hash.toHex() == txHash.toHex()
  })
}

export function transactionByAppId(block: SignedBlock, appId: number): GenericExtrinsic[] {
  return block.block.extrinsics.filter((tx) => {
    return extractAppIdFromTx(tx) == appId
  })
}

export function transactionHashToIndex(block: SignedBlock, txHash: H256): number[] {
  const indices: number[] = []
  for (const [index, tx] of block.block.extrinsics.entries()) {
    if (tx.hash.toHex() == txHash.toHex()) {
      indices.push(index)
    }
  }
  return indices
}

export function dataSubmissionsCount(block: SignedBlock): number {
  return dataSubmissionsAll(block).length
}

export function dataSubmissionsAll(block: SignedBlock): DataSubmission[] {
  const dataSubmissions = []
  const txs = block.block.extrinsics
  for (let i = 0; i < txs.length; i += 1) {
    const ds = dataSubmissionsByIndex(block, i)
    if (ds != null) {
      dataSubmissions.push(ds)
    }
  }

  return dataSubmissions
}

export function dataSubmissionsBySigner(block: SignedBlock, signer: string): DataSubmission[] {
  return dataSubmissionsAll(block).filter((sub) => {
    return sub.txSigner == signer
  })
}

export function dataSubmissionsByHash(block: SignedBlock, txHash: H256): DataSubmission | null {
  const index = block.block.extrinsics.findIndex((tx) => {
    return tx.hash.toHex() == txHash.toHex()
  })

  if (index == -1) return null

  return dataSubmissionsByIndex(block, index)
}

export function dataSubmissionsByIndex(block: SignedBlock, txIndex: number): DataSubmission | null {
  const transactions = block.block.extrinsics
  if (txIndex >= transactions.length) return null

  const tx = block.block.extrinsics[txIndex]
  return extractDataSubmissionFromTx(tx, txIndex)
}

export function dataSubmissionsByAppId(block: SignedBlock, appId: number): DataSubmission[] {
  return dataSubmissionsAll(block).filter((sub) => {
    return sub.appId == appId
  })
}

export function extractDataSubmissionDataFromTx(tx: GenericExtrinsic): string | null {
  if (tx.method.section != "dataAvailability" && tx.method.method != "submitData") {
    return null
  }

  let dataHex = tx.method.args.map((a) => a.toString()).join(", ")
  if (dataHex.startsWith("0x")) {
    dataHex = dataHex.slice(2)
  }

  return dataHex
}

export function extractDataSubmissionFromTx(tx: GenericExtrinsic, txIndex: number): DataSubmission | null {
  const data = extractDataSubmissionDataFromTx(tx)
  if (data == null) return null

  const txHash = tx.hash
  const txSigner = tx.signer.toString()
  const appId = extractAppIdFromTx(tx)
  return new DataSubmission(txHash, txIndex, data, txSigner, appId)
}

export function extractAppIdFromTx(tx: GenericExtrinsic): number {
  return parseInt((tx as any).__internal__raw.signature.appId.toString())
}

export class DataSubmission {
  constructor(
    public txHash: H256,
    public txIndex: number,
    public hexData: string,
    public txSigner: string,
    public appId: number,
  ) {}

  static fromGenericTx(tx: GenericExtrinsic, txIndex: number): DataSubmission | null {
    return extractDataSubmissionFromTx(tx, txIndex)
  }

  toAscii(): string {
    return fromHexToAscii(this.hexData)
  }
}

File: avail-rust/src/block.rs
use crate::error::ClientError;
use crate::rpcs;
use crate::{
	avail::data_availability::calls::types as DataAvailabilityCalls,
	primitives::block::extrinsics_params::CheckAppId,
};
use crate::{
	ABlock, AExtrinsicDetails, AExtrinsicEvents, AExtrinsics, AFoundExtrinsic, AOnlineClient,
};

use primitive_types::H256;
use subxt::backend::rpc::reconnecting_rpc_client::RpcClient;
use subxt::backend::StreamOfResults;
use subxt::blocks::StaticExtrinsic;
use subxt::storage::StorageKeyValuePair;
use subxt::utils::Yes;

pub struct Block {
	pub block: ABlock,
	pub transactions: AExtrinsics,
}

impl Block {
	pub async fn new(client: &AOnlineClient, block_hash: H256) -> Result<Self, ClientError> {
		let (block, transactions) = fetch_transactions(client, block_hash).await?;
		Ok(Self {
			block,
			transactions,
		})
	}

	pub async fn new_best_block(
		online_client: &AOnlineClient,
		rpc_client: &RpcClient,
	) -> Result<Self, ClientError> {
		let block_hash = Self::fetch_best_block_hash(rpc_client).await?;
		Self::new(online_client, block_hash).await
	}

	pub async fn new_finalized_block(
		online_client: &AOnlineClient,
		rpc_client: &RpcClient,
	) -> Result<Self, ClientError> {
		let block_hash = Self::fetch_finalized_block_hash(rpc_client).await?;
		Self::new(online_client, block_hash).await
	}

	pub async fn from_block(block: ABlock) -> Result<Self, subxt::Error> {
		let transactions = block.extrinsics().await?;
		Ok(Self {
			block,
			transactions,
		})
	}

	pub async fn from_block_number(
		online_client: &AOnlineClient,
		rpc_client: &RpcClient,
		block_number: u32,
	) -> Result<Self, ClientError> {
		let block_hash = rpcs::get_block_hash(rpc_client, Some(block_number)).await?;
		Self::new(online_client, block_hash).await
	}

	pub async fn events(
		&self,
		tx_index: Option<u32>,
	) -> Result<Vec<AExtrinsicEvents>, ClientError> {
		fetch_events(&self.transactions, tx_index).await
	}

	pub fn transaction_all_static<T: StaticExtrinsic>(&self) -> Vec<AFoundExtrinsic<T>> {
		transaction_all_static::<T>(&self.transactions)
	}

	pub fn transaction_count(&self) -> usize {
		transaction_count(&self.transactions)
	}

	pub fn transaction_by_signer(&self, signer: &str) -> Vec<AExtrinsicDetails> {
		transaction_by_signer(&self.transactions, signer)
	}

	pub fn transaction_by_signer_static<T: StaticExtrinsic>(
		&self,
		signer: &str,
	) -> Vec<AFoundExtrinsic<T>> {
		transaction_by_signer_static::<T>(&self.transactions, signer)
	}

	pub fn transaction_by_index(&self, tx_index: u32) -> Option<AExtrinsicDetails> {
		transaction_by_index(&self.transactions, tx_index)
	}

	pub fn transaction_by_index_static<T: StaticExtrinsic>(
		&self,
		tx_index: u32,
	) -> Option<AFoundExtrinsic<T>> {
		transaction_by_index_static::<T>(&self.transactions, tx_index)
	}

	pub fn transaction_by_hash(&self, tx_hash: H256) -> Vec<AExtrinsicDetails> {
		transaction_by_hash(&self.transactions, tx_hash)
	}

	pub fn transaction_by_hash_static<T: StaticExtrinsic>(
		&self,
		tx_hash: H256,
	) -> Vec<AFoundExtrinsic<T>> {
		transaction_by_hash_static::<T>(&self.transactions, tx_hash)
	}

	pub fn transaction_by_app_id(&self, app_id: u32) -> Vec<AExtrinsicDetails> {
		transaction_by_app_id(&self.transactions, app_id)
	}

	pub fn transaction_by_app_id_static<T: StaticExtrinsic>(
		&self,
		app_id: u32,
	) -> Vec<AFoundExtrinsic<T>> {
		transaction_by_app_id_static(&self.transactions, app_id)
	}

	pub fn transaction_hash_to_index(&self, tx_hash: H256) -> Vec<u32> {
		transaction_hash_to_index(&self.transactions, tx_hash)
	}

	pub fn data_submissions_all(&self) -> Vec<DataSubmission> {
		data_submissions_all(&self.transactions)
	}

	pub fn data_submissions_by_signer(&self, signer: &str) -> Vec<DataSubmission> {
		data_submissions_by_signer(&self.transactions, signer)
	}

	pub fn data_submissions_by_index(&self, tx_index: u32) -> Option<DataSubmission> {
		data_submissions_by_index(&self.transactions, tx_index)
	}

	pub fn data_submissions_by_hash(&self, tx_hash: H256) -> Option<DataSubmission> {
		data_submissions_by_hash(&self.transactions, tx_hash)
	}

	pub fn data_submissions_by_app_id(&self, app_id: u32) -> Option<DataSubmission> {
		data_submissions_by_app_id(&self.transactions, app_id)
	}

	pub async fn storage_fetch<'address, T>(
		&self,
		address: &'address T,
	) -> Result<Option<<T as subxt::storage::Address>::Target>, subxt::Error>
	where
		T: subxt::storage::Address<IsFetchable = Yes> + 'address,
	{
		self.block.storage().fetch(address).await
	}

	pub async fn storage_fetch_or_default<'address, T>(
		&self,
		address: &'address T,
	) -> Result<<T as subxt::storage::Address>::Target, subxt::Error>
	where
		T: subxt::storage::Address<IsFetchable = Yes, IsDefaultable = Yes> + 'address,
	{
		self.block.storage().fetch_or_default(address).await
	}

	pub async fn storage_iter<T>(
		&self,
		address: T,
	) -> Result<StreamOfResults<StorageKeyValuePair<T>>, subxt::Error>
	where
		T: subxt::storage::Address<IsIterable = Yes> + 'static,
		T::Keys: 'static + Sized,
	{
		self.block.storage().iter(address).await
	}

	pub async fn fetch_best_block_hash(client: &RpcClient) -> Result<H256, ClientError> {
		rpcs::get_block_hash(client, None).await
	}

	pub async fn fetch_finalized_block_hash(client: &RpcClient) -> Result<H256, ClientError> {
		rpcs::get_finalized_head(client).await
	}
}

pub async fn fetch_transactions(
	client: &AOnlineClient,
	block_hash: H256,
) -> Result<(ABlock, AExtrinsics), subxt::Error> {
	let block = client.blocks().at(block_hash).await?;
	let transactions = block.extrinsics().await?;
	Ok((block, transactions))
}

pub async fn fetch_events(
	transactions: &AExtrinsics,
	tx_index: Option<u32>,
) -> Result<Vec<AExtrinsicEvents>, ClientError> {
	let mut events = Vec::new();
	for details in transactions
		.iter()
		.filter(|tx| tx_index.map_or(true, |index| details.index() == index))
	{
		events.push(details.events().await?);
	}
	Ok(events)
}

pub fn transaction_count(transactions: &AExtrinsics) -> usize {
	return transactions.len();
}

pub fn transaction_all_static<T: StaticExtrinsic>(
	transactions: &AExtrinsics,
) -> Vec<AFoundExtrinsic<T>> {
	transactions.find::<T>().flatten().collect()
}

pub fn data_submissions_all(transactions: &AExtrinsics) -> Vec<DataSubmission> {
	transaction_all_static::<DataAvailabilityCalls::SubmitData>(transactions)
		.into_iter()
		.map(|tx| DataSubmission::from_static(tx))
		.collect()
}

pub fn transaction_by_signer(transactions: &AExtrinsics, signer: &str) -> Vec<AExtrinsicDetails> {
	transactions
		.iter()
		.filter(|tx| tx.signature_bytes() == Some(signer.as_bytes()))
		.collect()
}

pub fn transaction_by_signer_static<T: StaticExtrinsic>(
	transactions: &AExtrinsics,
	signer: &str,
) -> Vec<AFoundExtrinsic<T>> {
	transactions
		.find::<T>()
		.flatten()
		.filter(|tx| tx.details.signature_bytes() == Some(signer.as_bytes()))
		.collect()
}

pub fn data_submissions_by_signer(transactions: &AExtrinsics, signer: &str) -> Vec<DataSubmission> {
	transaction_by_signer_static::<DataAvailabilityCalls::SubmitData>(transactions, signer)
		.into_iter()
		.map(|tx| DataSubmission::from_static(tx))
		.collect()
}

pub fn transaction_by_index(
	transactions: &AExtrinsics,
	tx_index: u32,
) -> Option<AExtrinsicDetails> {
	transactions.iter().find(|tx| tx.index() == tx_index)
}

pub fn transaction_by_index_static<T: StaticExtrinsic>(
	transactions: &AExtrinsics,
	tx_index: u32,
) -> Option<AFoundExtrinsic<T>> {
	let details = transactions.iter().find(|tx| tx.index() == tx_index)?;
	let value = details.as_extrinsic::<T>().ok()??;

	Some(AFoundExtrinsic { details, value })
}

pub fn data_submissions_by_index(
	transactions: &AExtrinsics,
	tx_index: u32,
) -> Option<DataSubmission> {
	transaction_by_index_static::<DataAvailabilityCalls::SubmitData>(transactions, tx_index)
		.map(|tx| DataSubmission::from_static(tx))
}

pub fn transaction_by_hash(transactions: &AExtrinsics, tx_hash: H256) -> Vec<AExtrinsicDetails> {
	transactions
		.iter()
		.filter(|tx| tx.hash() == tx_hash)
		.collect()
}

pub fn transaction_by_hash_static<T: StaticExtrinsic>(
	transactions: &AExtrinsics,
	tx_hash: H256,
) -> Vec<AFoundExtrinsic<T>> {
	transactions
		.find::<T>()
		.flatten()
		.filter(|tx| tx.details.hash() == tx_hash)
		.collect()
}

pub fn data_submissions_by_hash(
	transactions: &AExtrinsics,
	tx_hash: H256,
) -> Option<DataSubmission> {
	transaction_by_hash_static::<DataAvailabilityCalls::SubmitData>(transactions, tx_hash)
		.into_iter()
		.map(|tx| DataSubmission::from_static(tx))
		.next()
}

pub fn transaction_by_app_id(transactions: &AExtrinsics, app_id: u32) -> Vec<AExtrinsicDetails> {
	transactions
		.iter()
		.filter(|tx| read_app_id(tx) == Some(app_id))
		.collect()
}

pub fn transaction_by_app_id_static<T: StaticExtrinsic>(
	transactions: &AExtrinsics,
	app_id: u32,
) -> Vec<AFoundExtrinsic<T>> {
	transactions
		.find::<T>()
		.flatten()
		.filter(|tx| read_app_id(&tx.details) == Some(app_id))
		.collect()
}

pub fn data_submissions_by_app_id(
	transactions: &AExtrinsics,
	app_id: u32,
) -> Option<DataSubmission> {
	let all_submissions: Vec<DataSubmission> =
		transaction_by_app_id_static::<DataAvailabilityCalls::SubmitData>(transactions, app_id)
			.into_iter()
			.map(|tx| DataSubmission::from_static(tx))
			.collect();

	all_submissions.into_iter().next()
}

pub fn transaction_hash_to_index(transactions: &AExtrinsics, tx_hash: H256) -> Vec<u32> {
	let mut indices = Vec::new();
	for tx in transactions.iter() {
		if tx.hash() == tx_hash {
			indices.push(tx.index())
		}
	}

	return indices;
}

pub fn read_app_id(transaction: &AExtrinsicDetails) -> Option<u32> {
	transaction
		.signed_extensions()?
		.find::<CheckAppId>()
		.ok()?
		.and_then(|e| Some(e.0))
}

#[derive(Debug, Clone)]
pub struct DataSubmission {
	pub tx_hash: H256,
	pub tx_index: u32,
	pub data: Vec<u8>,
	pub tx_signer: Vec<u8>,
	pub app_id: u32,
}

impl DataSubmission {
	pub fn from_static(tx: AFoundExtrinsic<DataAvailabilityCalls::SubmitData>) -> Self {
		let tx_hash = tx.details.hash();
		let tx_index = tx.details.index();
		let tx_signer = tx
			.details
			.signature_bytes()
			.expect("DA can only be executed signed")
			.to_vec();
		let app_id = read_app_id(&tx.details).expect("There must be an app id");
		let data = tx.value.data.0.clone();
		Self {
			tx_hash,
			tx_index,
			data,
			tx_signer,
			app_id,
		}
	}

	pub fn to_ascii(&self) -> Option<String> {
		to_ascii(self.data.clone())
	}
}

pub fn to_ascii(value: Vec<u8>) -> Option<String> {
	String::from_utf8(value).ok()
}
